### DISCUSSION TITLE
Summary of modularization oslo effort that will go into noresm2_3 release

### DISCUSSION THREAD

**Discussion by mvertens** [NorESMhub/CAM#140] [General]
**What motivated the following?:**

 - create new oslo_aero repository with
   - **src/** directory for main oslo_aero code
   - **src_cam/** directory for shadow files  
- no more common blocks
- no more #ifdefs
- minimize number of separate files
  - combine in one module 	 
  - also move to new private routines in modules if they are just called in one place
- long term goal - would want to have src directory  be able to be used in multiple host models

**new: $SRCROOT/components/cam/src/chemistry/oslo_aero/src**                        

- aero_model.F90
- oslo_aero_aerocom.F90
- oslo_aero_aerocom_tables.F90
- oslo_aero_aerodry_tables.F90
- oslo_aero_coag.F90
- oslo_aero_conc.F90
- oslo_aero_condtend.F90
- oslo_aero_control.F90
- oslo_aero_depos.F90
- oslo_aero_diurnal_var.F90
- oslo_aero_dust.F90
- oslo_aero_dust_sediment.F90
- oslo_aero_hetfrz.F90
- oslo_aero_linear_interp.F90
- oslo_aero_logn_tables.F90
- oslo_aero_microp.F90
- oslo_aero_ndrop.F90
- oslo_aero_nucleate_ice.F90
- oslo_aero_ocean.F90
- oslo_aero_optical_params.F90
- oslo_aero_seasalt.F90
- oslo_aero_share.F90
- oslo_aero_sox_cldaero.F90
- oslo_aero_sw_tables.F90

**new: $SRCROOT/components/cam/src/chemistry/oslo_aero/src_cam**

- chemistry.F90
- micro_mg_data.F90
- mo_chm_diags.F90
- mo_drydep.F90
- mo_gas_phase_chemdr.F90
- mo_neu_wetdep.F90
- mo_photo.F90
- mo_setaer.F90
- mo_setsox.F90
- mo_srf_emissions.F90
- mo_usrrxt.F90
- radiation.F90
- radlw.F90
- radsw.F90
- runtime_opts.F90
- vertical_diffusion.F90
- zm_microphysics.F90

**old: $SRCROOT/components/cam/src/cam/src/physics/cam_oslo**

- aerocopt.h
  removed and data moved to oslo_aero_aerocom_tables.F90

- aerocopt2.h
  removed and data moved to oslo_aero_aerocom_tables.F90

- aerodry.h
  removed and data moved to oslo_aero_aerodry_tables.F90

- checkTableHeader.F90
  removed and private routines in oslo_aero_aerocom_tables.F90, oslo_aero_aerodry_tables.F90, oslo_aero_logn_tables.F90 and oslo_aero_sw_tables.F90

- coltst4intcons.F90
  removed since only needed if COLTSTINTCONS was defined, and this is no longer the case

- initaeropt.F90
  file renamed to oslo_aero_aerocom_tables.F90
  read in the AeroCom look-up tables for aerosol optical properties. 

- initdryp.F90
  file renamed to oslo_aero_aerodry_tables.F90
  read in the AeroCom look-up tables for calculating dry aerosol size and mass distribution properties.

- inputForInterpol.F90
  removed as a stand-alone file - now a private routine in oslo_aero_optical_params.F90 

- intaeropt0.F90, intaeropt1.F90, intaeropt2to3.F90, intaeropt4.F90, intaeropt5to10.F90
  removed as stand-alone files - now routines in oslo_aero_aerocom_tables.F90 and called by oslo_aero_aerocom.F90  

- intdrypar0.F90, intdrypar1.F90, intdrypar2to3.F90, intdrypar4.F90, intdrypar5to10.F90
  removed as stand-alone files - now routines in oslo_aero_aerocom_tables.F90 and called by oslo_aero_aerocom.F90

- intfrh.F90
  removed - not used 

- lininterpol3dim.F90, lininterpol4dim.F90, lininterpol5dim.F90
  removed as stand-alone routines - now public routine in oslo_aero_linear_interp

- opticsAtConstRh.F90
  removed as a stand-alone routine - now a private routine in  oslo_aero_aerocom.F90

- opttab.F90, opttab_lw.F90, optinterpol.F90 
  removed as stand-alone routiens and combined in new routine module,  oslo_aero_sw_tables.F90

- oslo_control.F90
  just formatting changes
  
- pmxsub.F90
  renamed oslo_aero_optical_parameters.F90 and refactored to leverage the following
  use oslo_aero_share,     only: eps, rh, fombg, fbcbg, fac, fbc, faq, cate, cat
  use oslo_aero_share,     only: nmodes, nbmodes, nbands, nlwbands
  use oslo_aero_share,     only: rhopart, l_soa_na, l_so4_na
  use oslo_aero_share,     only: calculateNumberConcentration
  use oslo_aero_conc,      only: calculateBulkProperties, partitionMass
  use oslo_aero_sw_tables, only: interpol0, interpol1, interpol2to3, interpol4, interpol5to10
  use oslo_aero_aerocom,   only: aerocom1, aerocom2

- ptaero_table.F90, table_manager.F90
  removed - not used
  
- radiation.F90
  shadow file: refactored and moved to oslo_aero/src_cam

- radlw.F90
  shadow file: refactored and moved to oslo_aero/src_cam 

- radsw.F90
  shadow file: refactored and moved to oslo_aero/src_cam 

**old: $SRCROOT/components/cam/src/cam/src/chemistry/oslo_aero**

- aero_model.F90, constants.F90
  constants.F90 removed and incorporated into aero_model.F90

- aerosoldef.F90, commondefinitions.F90, const.F90, modal_aero_data.F90, oslo_utils.F90
  removed stand-alone routines and incorporated into new module oslo_aero_share.F90

- aeronucl.F90, appformrate.F90
  removed as stand-alone routines and made a private routine in oslo_aero_condtend.F90
  
- calcaersize.F90
  removed as a stand-alone routine and made a private routine in aero_model.F90

- condtend.F90
  renamed oslo_aero_contend.F90 and refactored
  
- dust_model.F90
  renamed as oslo_aero_dust.F90 and refactored

- dust_sediment_mod.F90
  renamed  as oslo_aero_dust_sediment.F90 and called by oslo_aero_depos.F90

- hetfrz_classnuc_oslo.F90
  renamed as oslo_aero_hetfrz.F90 and refactored

- initlogn.F90, intlog1to3.F90, intlog4.F90, intlog5to10.F90
  removed as stand-alone files and incorporated into new file oslo_aero_logn_tables.F90

- koagsub.F90
  rename oslo_aero_coag.F90 and refactored
  
- microp_aero.F90
  renamed oslo_aero_microp.F90 *** this might be a problem in noresm2_5_develop ***	

- modal_aero_deposition.F90, oslo_aerosols_intr.F90
  renamed oslo_aero_depos.F90 and refactored *** this might be a problem in noresm2_5_develop *** 
  oslo_aerosols_intr.F90 - removed as a stand-alone routine and incporated into oslo_aero_depos.F90 

- parmix_progncdnc.F90, modalapp2d.F90
  removed as stand-alones routines and merged into new routine oslo_aero_conc.F90

- ndrop.F90
  renamed oslo_aero_ndrop.F90 and significantly rewritten
  public  :: ndrop_init_oslo
  public  :: dropmixnuc_oslo
  private :: explmix_oslo
  private :: maxsat_oslo
  private :: ccncalc_oslo
  private :: activate_modal_oslo

- nucleate_ice_oslo.F90
  renamed oslo_aero_nucleate_ice.f90

- oslo_ocean_intr.F90
   rename to oslo_aero_ocean.F90
   
- seasalt_model.F90
  renamed oslo_aero_seasalt.F90

- sox_cldaero_mod.F90
  renamed oslo_aero_sox_cldaero
  
- Moved to shadow files in src_cam/
  mo_chm_diags.F90
  mo_drydep.F90
  mo_extfrc.F90
  mo_gas_phase_chemdr.F90
  mo_neu_wetdep.F90
  mo_setsox.F90
  mo_srf_emissions.F90
  mo_usrrxt.F90
  vertical_diffusion.F90
  zm_microphysics.F90
  oxi_diurnal_var.F90 (now part of shadow file mo_gas_phase_chemdr.F90)

**Comment by DirkOlivie** [comment]
I have a question concerning the subroutine surf_area_dens that was defined in oslo_aero before (NorESM2.1), but not anymore in NorESM2.3.

The subroutine surf_area_dens is called in oslo_aero/src/aero_model.F90, but is not anymore defined in aero_model.F90 (it was defined there in NorESM2.1).  It however still exists in chemistry/modal_aero/aero_model.F90.  Does it mean that the code in chemistry/modal_aero is also used/compiled when running oslo_aero?

**Reply by gold2718** [reply]
Are you planning on using the NorESM 2.1 version? If so, we can restore it. It was a stub (no calculations) with a slightly different calling list.

**Comment by mvertens** [comment]
@DirkOlivie
- in noresm2.3 - the subroutine surf_area_dens is not called from oslo_aero/src/aero_model.F90. It does not exist at all in the new oslo_aero repository. I'm not sure where you are seeing this call. 
- in noresm2.1 - the only call to surf_area_dens was in chemistry_modal/aero_model.F90 - which is not used when you turn on oslo_aero. The suroutine is in fact in chemistry/oslo_aerp/aero_model.F90 and in chemistry/modal_aero/aero_model.F90- but is not used and therefore is not needed when running oslo_aero. That was the reason it was removed in noresm2.3.

**Reply by DirkOlivie** [reply]
@mvertens @gold2718 
Thanks for the clarification.  In the full-chemistry version, there will be a call to surf_area_dens.  So if it is ok, I will bring it in again (will be part of my pull request). 
It is not the same code as in chemistry/modal_aero/aero_model.F90, but adapted to the oslo_aero aerosols.

**Reply by gold2718** [reply]
:+1: 
Please just ensure that the calling interface is identical to the one in CAM.
